### PR TITLE
Allen Russell

### PR DESCRIPTION
**Pull request by garussell**
I thought this IC went fairly well with a few roadblocks.  I think I confused myself because there was not a project index and no contestant show page in the IC.  I created a project index with a link to the project show page which was probably unnecessary.  

One thing that tripped me up was where to put the logic for 'project_count'.  So I created a contestant_projects controller using the terminal command, `rails generate controller ContestantProjects` and that created files in the `helpers` and  `requests` folders that I subsequently deleted.  I realize now it wasn't needed for what was being asked, and learned from it.  

I time-boxed this IC and did not spend a minute over 3-hours as requested.  I took regular Poms and got through 3/3 user stories.  Although I was tempted to add more to the project than what was required, getting a bit ahead of myself, I appreciate the practice this IC provided.

### CODE REVIEW

File: app/controllers/contestants_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class ContestantsController < ApplicationController
+  def index
+    @contestants = Contestant.includes(:projects)
```
Comment (cjsim89) [inline app/controllers/contestants_controller.rb:3]: Nice job using `.includes` here - we may or may not have gone over this in class. Just be sure you know what this method is doing underneath, and why you are using it instead of the more basic `.all` in this controller!

File: app/controllers/projects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class ProjectsController < ApplicationController
+  def index
+    @project = Project.all
+  end
+
+  def show
+    @project = Project.all
```
Comment (cjsim89) [inline app/controllers/projects_controller.rb:7]: For a show page, we want to focus on only ONE instance at a time. This query gives you ALL instances of Projects, and I see from your view that you're looping through all of them. I think it might be good to review what each action here does, and what they correspond to, specifically index and show: 

* Index is used as a _big list_ of all the instances of a given resource (in this case, Projects). 
* Show is used as a _detail_ page for one specific instance of a given resource (in this case, Project [singular]). 

in other words, an index is the entire Recipe Box, and a show is Meatloaf.

Comment (cjsim89) [reply app/controllers/projects_controller.rb:7]: The correct query for the show action would be: 
```
@project = Project.find(params[:id])
```

To check this yourself, put a `binding.pry` on line 6.5 and see what `params` is (should be an `id` field in there). That ID is referencing one specific instance of a project, and that's the one we want to see (the _only_ one).

File: app/models/project.rb [hunk 3]
```diff
@@ -1,4 +1,11 @@
 class Project < ApplicationRecord
   validates_presence_of :name, :material
   belongs_to :challenge
+  has_many :contestant_projects
+  has_many :contestants, through: :contestant_projects
+
+  def project_count
+    contestant_projects.count
```
Comment (cjsim89) [inline app/models/project.rb:8]: Great job writing this query in the model!

File: app/views/projects/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<h1>Home Page</h1>
```
Comment (cjsim89) [inline app/views/projects/index.html.erb:1]: So, an index page isn't necessarily "the home page" - as I wrote above, the index page for a given resource is usually the big list of all the instances of that resource. 

This page wasn't required by the user stories, so next time I would leave it out unless you are practicing something specific.

File: app/views/projects/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+  <%= link_to "Projects", "/projects/show" %><br>
```
Comment (cjsim89) [inline app/views/projects/index.html.erb:2]: This link is also misleading since a show page will require a specific Id... Going to check on your routes...

Comment (cjsim89) [reply app/views/projects/index.html.erb:2]: I see now - your routes don't have a path to `projects/show`; but there is a controller-action pairing that would work. If you didn't try it, run `rails s` and see if you can click this link - it shouldn't work because we want to put the exact path (URL) to a specific project. It would be something like `/projects/2` or whatever ID exists in the database for a specific project instance.

File: app/views/projects/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<h1>Projects</h1>
+
+<% @project.each do |project| %>
```
Comment (cjsim89) [inline app/views/projects/show.html.erb:3]: Overall this set up is good, but you _should_ only have one instance of `@project` here, and we wouldn't need to loop through it using `.each`. Otherwise, your page is set up correctly.

File: config/routes.rb [hunk 7]
```diff
@@ -2,5 +2,9 @@
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
   # Defines the root path route ("/")
-  # root "articles#index"
+  root "projects#index"
```
Comment (cjsim89) [inline config/routes.rb:5]: Ohhhh I see now! this makes sense as a homepage. 👍 
Generally this isn't required on an IC, we'll just know to go to whatever pages are requested by each user story, so you can skip this part in the future (if you want).

File: spec/features/contestants/index_spec.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    @rug_tux = @furniture_challenge.projects.create!(name: "Rug Tuxedo", material: "Rug")
+    @lit_fit = @furniture_challenge.projects.create!(name: "Litfit", material: "Lamp")
+    @leather_feather = @furniture_challenge.projects.create!(name: "LeatherFeather", material: "Leather")
+    
+    @kentaro = Contestant.create!(name: "Kentaro Kameyama", age: 30, hometown: "Boston", years_of_experience: 8)
+    @jay = Contestant.create!(name: "Jay McCarroll", age: 40, hometown: "LA", years_of_experience: 13)
+  
+    ContestantProject.create!(contestant: @kentaro, project: @lit_fit)
+    ContestantProject.create!(contestant: @kentaro, project: @rug_tux)
+    ContestantProject.create!(contestant: @jay, project: @leather_feather)
```
Comment (cjsim89) [inline spec/features/contestants/index_spec.rb:16]: Good job with your object setup!

File: spec/features/projects/index_spec.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    @furniture_challenge = Challenge.create!(theme: "Apartment Furnishings", project_budget: 1000)
+      
+    @lit_fit = @furniture_challenge.projects.create!(name: "Litfit", material: "Lamp")
+    @leather_feather = @furniture_challenge.projects.create!(name: "LeatherFeather", material: "Leather")
+  end
+  describe "when I visit the root page" do
+    it "has a link to each project" do
+      visit "/projects"
+
+      expect(page).to have_link("Projects")
```
Comment (cjsim89) [inline spec/features/projects/index_spec.rb:14]: This test is ok, you may want to include an `href` key with the value of where exactly this page goes... E.g.
```
expect(page).to have_link("Projects", href: "/projects")
```

File: spec/features/projects/show_spec.rb [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+  describe "as a visitor" do
+    describe "when I visit a project show page" do
+      it "I see the theme of the challenge and projects they belong to" do
+
+        visit "/projects/#{@lit_fit.id}"
+
+        expect(page).to have_content("Litfit")
+        expect(page).to have_content("Lamp")
+        expect(page).to have_content("Apartment Furnishings")
```
Comment (cjsim89) [inline spec/features/projects/show_spec.rb:33]: This test is also ok, but I think a closer reading of the user story may have led you down the path of only having one instance on this paeg instead of many; the test doesn't really cover this either. Instead, if you are only visiting the `lit_fit` show page, you shouldn't see any content related to a different project. E.g. 

```
expect(page).to_not have_content(@leather_feather.name)
```

Currently, if that line was in this test, it would fail - which would prove that it's not covering the user story entirely accurately. That line is also missing from your `it` block, so be careful when writing your describes and it blocks!

File: spec/models/project_spec.rb [hunk 11]
```diff
@@ -11,4 +11,23 @@
...
+      
+      kentaro = Contestant.create!(name: "Kentaro Kameyama", age: 30, hometown: "Boston", years_of_experience: 8)
+      jay = Contestant.create!(name: "Jay McCarroll", age: 40, hometown: "LA", years_of_experience: 13)
+    
+      ContestantProject.create!(contestant: kentaro, project: lit_fit)
+      ContestantProject.create!(contestant: kentaro, project: leather_feather)
+      ContestantProject.create!(contestant: jay, project: leather_feather)
+
+      expect(leather_feather.project_count).to eq(2)
+      expect(lit_fit.project_count).to eq(1)
```
Comment (cjsim89) [inline spec/models/project_spec.rb:30]: Good job with this model test!

File: spec/features/contestants/index_spec.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+  describe "as a visitor" do
+    describe "when I visit the contestants index page" do
+      it "has a list of names of all the contestants and projects they are assigned to" do
+
+        visit "/contestants"
+
+        expect(page).to have_content("Kentaro Kameyama")
+        expect(page).to have_content(@lit_fit.name)
+        expect(page).to have_content(@rug_tux.name)
```
Comment (cjsim89) [inline spec/features/contestants/index_spec.rb:27]: In the future, this test can be improved with the addition of the `within` block - otherwise, how can we tell that LIt Fit and Rug Tux belong to Kentaro and NOT Jay? 
If we tie them together on the page using some specific ids, we'd have a surefire way to know that the page is constructed correctly with the right projects under the right contestants.

Comment (cjsim89) [reply spec/features/contestants/index_spec.rb:27]: If your view looked something like this: 
```
{ contestant info here } 
<ul id="contestant_projects">
   <% @contestant.projects.each do |project| %>
       <li id="project_<%= project.id %>"><%= project.name %></li>
   <% end %>
</ul>
```
...then we can tie the projects to the contestant info that's being looped through above. Try it out and let me know what you think!

Comment (cjsim89) [reply spec/features/contestants/index_spec.rb:27]: The corresopnding test might look like: 

```
within "#contestant_#{id}" do
   within "#project_2" do
       expect(page).to have_content("this project")
       expect(page).to_not have_content("that other project")

   end

end
